src/main/java/de/tum/in/www1/artemis/service/competency/CompetencyProgressService.java

```diff
@@ -217,26 +204,187 @@ public CompetencyProgress updateCompetencyProgress(Long competencyId, User user)
+
+        double recencyConfidenceHeuristic = calculateRecencyConfidenceHeuristic(participantScoreInfos);
+        double difficultyConfidenceHeuristic = calculateDifficultyConfidenceHeuristic(participantScoreInfos, exerciseInfos);
+        double quickSolveConfidenceHeuristic = calculateQuickSolveConfidenceHeuristic(participantScoreInfos);
+
+        double confidence = 1 + recencyConfidenceHeuristic + difficultyConfidenceHeuristic + quickSolveConfidenceHeuristic;
+
+        competencyProgress.setConfidence(confidence);
+        setConfidenceReason(competencyProgress, recencyConfidenceHeuristic, difficultyConfidenceHeuristic, quickSolveConfidenceHeuristic);
     }
 
     /**
-     * Calculate the confidence score for the given user in a competency.
+     * Calculate the recency confidence heuristic for the given user in a competency based on the exercises linked to the competency.
+     * If the recent scores are higher than the average scores, the confidence should also be higher and vice versa.
      *
-     * @param exercises A list of all exercises linked to a specific competency
-     * @param user      The user for which the confidence score should be calculated
-     * @return The average score of the user in all exercises linked to the competency
+     * @param participantScores the participant scores for the exercises linked to the competency
+     * @return The recency confidence heuristic
      */
-    private double calculateConfidence(@NotNull Set<Exercise> exercises, @NotNull User user) {
-        return participantScoreService.getStudentAndTeamParticipationScoresAsDoubleStream(user, exercises).summaryStatistics().getAverage();
+    private double calculateRecencyConfidenceHeuristic(@NotNull Set<CompetencyExerciseMasteryCalculationDTO> participantScores) {
+        if (participantScores.size() < 3) {
+            return 0;
+        }
+
+        Instant earliestScoreDate = participantScores.stream().map(CompetencyExerciseMasteryCalculationDTO::lastModifiedDate).min(Instant::compareTo).get();
+        Instant latestScoreDate = participantScores.stream().map(CompetencyExerciseMasteryCalculationDTO::lastModifiedDate).max(Instant::compareTo).get();
+
+        double doubleWeightedScoreSum = participantScores.stream()
+                .mapToDouble(info -> info.lastScore() * toRelativeTime(earliestScoreDate, latestScoreDate, info.lastModifiedDate())).sum();
+        double weightSum = participantScores.stream().mapToDouble(info -> toRelativeTime(earliestScoreDate, latestScoreDate, info.lastModifiedDate())).sum();
+        double weightedAverageScore = doubleWeightedScoreSum / weightSum;
+
+        double averageScore = participantScores.stream().mapToDouble(CompetencyExerciseMasteryCalculationDTO::lastScore).average().orElse(0.0);
+
+        double recencyConfidence = weightedAverageScore - averageScore;
+
+        return Math.clamp(recencyConfidence, -0.25, 0.25);
```

Comment:
Could we add variables for the confidence bounds and re-use them for the other confidences aswell?

```suggestion
        final double MIN_CONFIDENCE = -0.25;
        final double MAX_CONFIDENCE = 0.25;
        return Math.clamp(recencyConfidence, MIN_CONFIDENCE, MAX_CONFIDENCE);
```